core/src/main/java/org/apache/calcite/rex/RexBuilder.java

```diff
@@ -1336,8 +1337,15 @@ protected RexLiteral makeLiteral(
       o = ((TimestampWithTimeZoneString) o).round(p);
       break;
     case DECIMAL:
-      if (o != null && type.getScale() != RelDataType.SCALE_NOT_SPECIFIED) {
-        assert o instanceof BigDecimal;
+      if (o == null) {
+        break;
+      }
+      assert o instanceof BigDecimal;
+      if (type instanceof IntervalSqlType) {
+        SqlIntervalQualifier qualifier = ((IntervalSqlType) type).getIntervalQualifier();
+        o = ((BigDecimal) o).multiply(qualifier.getUnit().multiplier);
```

Comment:
multiplier can also be fractional, can you also add a test for some of these units, like MILLISECONDS?
It may be easier to add tests to a quide file than to SqlOperatorTest.

Reply:
There are no intervals less than SECONDS. Multiplier for seconds is 1000.

Reply:
An analogous problem might happen with MONTH-YEAR intervals. (MONTH is the smallest unit, so would be analogous to SECONDS , and YEAR is larger, so would be analogous to MINUTE.) Can you add a test for that case?

Reply:
@julianhyde, test for MONTH has already been added with the last commit.